coffee-chats/src/main/java/com/google/step/coffee/entity/DateRange.java

```diff
@@ -0,0 +1,231 @@
+
+    if (start.after(end)) {
+      throw new InstantiationError("Start must be before end");
+    }
+  }
+
+  public Date getStart() {
+    return start;
+  }
+
+  public DateTime getDateTimeStart() {
+    return new DateTime(start);
+  }
+
+  public Date getEnd() {
+    return end;
+  }
+
+  public DateTime getDateTimeEnd() {
+    return new DateTime(end);
+  }
+
+  public Duration getDuration() {
+    return Duration.between(start.toInstant(), end.toInstant());
+  }
+
+  public TimePeriod toTimePeriod() {
+    return new TimePeriod().setStart(this.getDateTimeStart()).setEnd(this.getDateTimeEnd());
+  }
+
+  /**
+   * Combines two lists of corresponding start and end dates into a list of DateRange objects.
+   */
+  public static List<DateRange> combineLists(List<Date> startDates, List<Date> endDates)
+      throws InvalidEntityException {
+    List<DateRange> dateRanges = new ArrayList<>();
+
+    for (int i = 0; i < startDates.size(); i++) {
+      if (startDates.get(i).before(endDates.get(i))) {
+        DateRange dateRange = new DateRange(startDates.get(i), endDates.get(i));
+        dateRanges.add(dateRange);
+      } else {
+        throw new InvalidEntityException("Start date must be before corresponding end date");
+      }
+    }
+
+    return dateRanges;
+  }
+
+  /**
+   * Converts TimePeriods returned by Google Calendar API to list of DateRanges.
+   */
+  public static List<DateRange> toDateRanges(List<TimePeriod> timePeriods) {
+    return timePeriods.stream().map(DateRange::new).collect(Collectors.toList());
+  }
+
+  /**
+   * Creates a new DateRange which combines the range covered by two given ranges.
+   */
+  public static DateRange getMergedRange(DateRange range1, DateRange range2) {
+    if (range1.overlaps(range2) || range1.adjacentTo(range2)) {
+      Date mergedStart = range1.getStart().before(range2.getStart()) ?
+          range1.getStart() :
+          range2.getStart();
+      Date mergedEnd = range1.getEnd().after(range2.getEnd()) ?
+          range1.getEnd() :
+          range2.getEnd();
+
+      return new DateRange(mergedStart, mergedEnd);
+    } else {
+      return DateRange.EMPTY;
+    }
+  }
+
+  /**
+   * Returns a DateRange that covers the intersection of two ranges.
+   */
+  public static DateRange getIntersectingRange(DateRange range1, DateRange range2) {
+    if (range1.overlaps(range2)) {
+      Date overlappingStart = range1.getStart().after(range2.getStart()) ?
+          range1.getStart() :
+          range2.getStart();
+
+      Date overlappingEnd = range1.getEnd().before(range2.getEnd()) ?
+          range1.getEnd() :
+          range2.getEnd();
+
+      return new DateRange(overlappingStart, overlappingEnd);
+    } else {
+      return DateRange.EMPTY;
+    }
+  }
+
+  @Override
+  public int compareTo(DateRange otherRange) {
+    return this.getStart().compareTo(otherRange.getStart());
+  }
+
+  @Override
+  public int hashCode() {
+    return this.start.hashCode() * this.end.hashCode();
```

Comment:
`return Objects.hash(this.start, this.end)`?

Reply:
Done